// @ethereum-waffle/compiler/dist/cjs/utils.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const fs_1 = __importDefault(require("fs"));
const path_1 = __importDefault(require("path"));
exports.readFileContent = (path) => fs_1.default.readFileSync(path).toString();
exports.isPositiveIntegerString = (value) => typeof value === 'string' && /^\d+$/.test(value);
exports.eventParseResultToArray = (eventResult) => Object.keys(eventResult)
    .filter(exports.isPositiveIntegerString)
    .map((key) => eventResult[key]);
exports.isWarningMessage = (error) => error.severity === 'warning';
exports.isFile = (filePath) => fs_1.default.existsSync(filePath) && fs_1.default.lstatSync(filePath).isFile();
exports.flattenObjectArray = (array) => array.reduce((accum, object) => Object.assign(accum, object), {});
exports.last = (array) => array[array.length - 1];
exports.deepCopy = (obj) => JSON.parse(JSON.stringify(obj));
exports.isDirectory = (directoryPath) => fs_1.default.existsSync(exports.relativePathToWorkingDir(directoryPath)) &&
    fs_1.default.statSync(exports.relativePathToWorkingDir(directoryPath)).isDirectory();
exports.relativePathToWorkingDir = (pathName) => path_1.default.resolve(pathName);
